crates/miden-lib/src/account/interface/mod.rs

```diff
@@ -0,0 +1,312 @@
+    // Internal value of the `RpoFalcon512` holds the storage offset where the private key is
+    // stored
+    RpoFalcon512(u8),
+    Custom(Vec<AccountProcedureInfo>),
+}
+
+impl AccountComponentInterface {
+    /// Creates a set of [AccountComponentInterface] instances, specifying components which were
+    /// used to create an account with the provided procedures array.
+    pub fn from_procedures(procedures: &[AccountProcedureInfo]) -> BTreeSet<Self> {
+        let rpo_falcon_procs = rpo_falcon_512_library()
+            .mast_forest()
+            .procedure_digests()
+            .collect::<Vec<Digest>>();
+
+        debug_assert!(rpo_falcon_procs.len() == 1);
+        let rpo_falcon_proc = rpo_falcon_procs[0];
+
+        let mut component_interface_set = BTreeSet::new();
+
+        if basic_wallet_library().mast_forest().procedure_digests().all(|proc_digest| {
+            procedures.iter().any(|proc_info| proc_info.mast_root() == &proc_digest)
+        }) {
+            component_interface_set.insert(AccountComponentInterface::BasicWallet);
+        }
+
+        if basic_fungible_faucet_library()
+            .mast_forest()
+            .procedure_digests()
+            .all(|proc_digest| {
+                procedures.iter().any(|proc_info| proc_info.mast_root() == &proc_digest)
+            })
+        {
+            component_interface_set.insert(AccountComponentInterface::BasicFungibleFaucet);
+        }
+
+        procedures.iter().for_each(|proc_info| {
+            if proc_info.mast_root() == &rpo_falcon_proc {
+                component_interface_set
+                    .insert(AccountComponentInterface::RpoFalcon512(proc_info.storage_offset()));
+            }
+        });
+
+        let custom_interface_procs = procedures
+            .iter()
+            .filter(|proc_digest| {
+                !basic_wallet_library()
+                    .mast_forest()
+                    .procedure_digests()
+                    .any(|wallet_proc_digest| &wallet_proc_digest == proc_digest.mast_root())
+                    && !basic_fungible_faucet_library()
+                        .mast_forest()
+                        .procedure_digests()
+                        .any(|faucet_proc_digest| &faucet_proc_digest == proc_digest.mast_root())
+                    && (&rpo_falcon_proc != proc_digest.mast_root())
+            })
+            .cloned()
+            .collect::<Vec<AccountProcedureInfo>>();
+
+        if !custom_interface_procs.is_empty() {
+            component_interface_set
+                .insert(AccountComponentInterface::Custom(custom_interface_procs));
+        }
+
+        component_interface_set
+    }
+
+    // probably it would be convenient to also have `from_components` constructor
+}
+
+/// An [`AccountInterface`]
+pub struct AccountInterface {
+    account_id: AccountId,
+    auth: Vec<AuthScheme>,
+    component_interfaces: BTreeSet<AccountComponentInterface>,
+}
+
+impl AccountInterface {
```

Comment:
I would probably add a constructor - something like:
```rust
impl AccountInterface {
    pub fn new(account_id: AccountId, code: &AccountCode) -> Self {
        ...
    }
}
```

Also, I'd probably expose some convenience methods like `account_type()`, `is_faucet()`, `is_public()` (i.e., similar to what we have on the `Account` struct).

Reply:
I'm not sure that we actually need the constructor, because essentially it will be a duplication of the code from the `From<Account>` implementation.  The main issue with the authentication scheme vector: to be able to build it we need to have a public key, which we can obtain from the storage. So to make constructor work we should either provide the public keys, or auth schemes themselves, or the whole Account Storage. So it looks like casting the `Account` to the `AccountInterface` is more simple and user-friendly way of creating it.

If we need to have a constructor anyway, how should we provide the information about public key?

Reply:
Oh, I see - we need account storage to construct the interface. This may be a problem as ideally we don't want to rely on the full account being loaded to construct the interface (the account may be very big either because the storage is big or the vault is big or both).

Not sure what's the best solution here is. Maybe OK to keep this as is for now and create a follow-up issue.

Reply:
For now I created a constructor which uses a vector of authentication schemes. Not sure that we will have them at the point of using the stateless check, but we at least don't need to have the account storage.